flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificTypeStrategies.java

```diff
@@ -58,8 +58,12 @@ public final class SpecificTypeStrategies {
     /** See {@link ArrayTypeStrategy}. */
     public static final TypeStrategy ARRAY = new ArrayTypeStrategy();
 
-    /** See {@link ArrayElementOutputTypeStrategy}. */
-    public static final TypeStrategy ARRAY_ELEMENT = new ArrayElementOutputTypeStrategy();
+    /** Type strategy specific for array element. */
+    public static final TypeStrategy ARRAY_ELEMENT =
+            callContext ->
+                    Optional.of(
```

Comment:
what if the passed type is not a `CollectionDataType`?

Reply:
@dawidwys The input will check whether it is array.  If not, it will throw exception

Reply:
@liuyongvs  I saw this class all like this    
```
/** See {@link ToTimestampLtzTypeStrategy}. */
    public static final TypeStrategy TO_TIMESTAMP_LTZ = new ToTimestampLtzTypeStrategy();
```

no one achieve class logic in this class, so why we don' t generate a class and new that class like others do?

Reply:
@hanyuzheng7 look at the map_keys map_values in this class  i supported before

Reply:
```
/** Specific {@link TypeStrategy} for {@link BuiltInFunctionDefinitions#ARRAY_MAX}. */
@Internal
public class ArrayElementOutputTypeStrategy implements TypeStrategy {
    @Override
    public Optional<DataType> inferType(CallContext callContext) {
        DataType inputDataType = callContext.getArgumentDataTypes().get(0);
        final DataType elementDataType = ((CollectionDataType) inputDataType).getElementDataType();
            return Optional.of(elementDataType.nullable());
    }
}
```

Reply:
I think this may have some problem.
https://github.com/apache/flink/pull/22730#discussion_r1242154811

Reply:
@hanyuzheng7 why has problem.  I removed this class

Reply:
this is you strategy
```
public  final TypeStrategy ARRAY_ELEMENT =
            callContext ->
                    Optional.of ((CollectionDataType) callContext.getArgumentDataTypes().get(0))
                                    .getElementDataType());
```
But you don't consider this case [ARRAY(INT NOT NULL) -> output: INT

Reply:
@hanyuzheng7 why my strategy answer is -1。my strategy is get arg0 's element type.  for your example is it not null.  Then force nullable to int

Reply:
oh i see, can you tell me where is the bug in this code?
flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayElementOutputTypeStrategy.java

Reply:
It seem the return in this class is always nullable. so why you change it?

Reply:
You may need look the issue talk. 
You should not set nullable according to input whether it is nullable 

because we don't know the value whether is null in compile procedure, only can know in the runtime.
for example:
ddl :

CREATE TABLE data_source (
  a array<int not null> not null
) WITH (
  'connector'='xxx',
);
// the element is array(), that is to say, empty array. the result is null.
// if the return type is int not null, how to save null value.
select array_max(a) from data_source;

Reply:
OK, I know that so you test case is empty()?

Reply:
@hanyuzheng7 It is not important. The important is result type should always be nullable.  So i add test change the array<int not null> not null to int
If your original code it will return int not null., which is not correct

Reply:
my original code return is int
you can see the test case
[flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/strategies/ArrayElementOutputTypeStrategyTest.java](url)

Reply:
Please test first @hanyuzheng7

Reply:
> @dawidwys The input will check whether it is array. If not, it will throw exception

I don't think so, this strategy named ARRAY_ELEMENT. you must to check it's input is array first. you cannot guarantee that others check it is array in input strategy.